### PR TITLE
New collumn of execution time added in db schema of oc_jobs

### PR DESCRIPTION
**Pull request by noveens**
## Description

I have changed the schema of the db_structure.xml file and the corresponding tests where oc_jobs was concerned.

I added a column of execution duration in the oc_jobs table.

## Related Issue

https://github.com/owncloud/core/issues/27137

## Motivation and Context

This is a new feature which could help in testing of the system.

## How Has This Been Tested?

I have tested this thoroughly, even edited one of the tests because of the database migration.

## Screenshots (if appropriate):
Not necessary.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by PVince81** [comment]
This will likely not be enough.

In past ownCloud versions, the XML file was the authority regarding database schemas.
Since master / 10.0, we should use migrations to make database changes.

Have a look how it's done here: https://github.com/owncloud/core/blob/master/core/Migrations/Version20170111103310.php#L9.
Some docs about migrations: https://github.com/owncloud/documentation/pull/2742 (unfinished as this is brand new).

Let me know if you need assistance.

Once this is done the next step is to adjust the background job runner to measure the execution time and put it there.

**Comment by noveens** [comment]
@PVince81 , thanks for the quick review, I'll look into it as soon as possible.
Thanks for the links to ducumentation as well.

**Comment by noveens** [comment]
@PVince81 , please check now, I completed making another column in oc_jobs table named execution time through migration.

File: tests/lib/DB/schemDiffData/core.xml [hunk 1]
```diff
@@ -732,6 +732,14 @@
 				<notnull>false</notnull>
 			</field>
 
+			<field>
```
Comment (DeepDiver1975) [inline tests/lib/DB/schemDiffData/core.xml]: not needed

File: tests/lib/DB/schemDiffData/unsigned.xml [hunk 2]
```diff
@@ -61,6 +61,14 @@
 				<notnull>false</notnull>
 			</field>
 
+			<field>
```
Comment (DeepDiver1975) [inline tests/lib/DB/schemDiffData/unsigned.xml]: not needed

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

File: tests/lib/DB/schemDiffData/unsigned.xml (hunk 2 above)
Comment (noveens) [inline tests/lib/DB/schemDiffData/unsigned.xml]: So I shouldn't add the new column in oc_jobs table in the schema of the tests?

Comment (DeepDiver1975) [inline tests/lib/DB/schemDiffData/unsigned.xml]: no this is not necessary

Comment (noveens) [inline tests/lib/DB/schemDiffData/unsigned.xml]: removing it.

**Comment by DeepDiver1975** [comment]
@noveens THx - next would be to fill the columns :wink:

**Comment by noveens** [comment]
@DeepDiver1975 , does that come in the scope of this bug?

**Comment by PVince81** [comment]
@noveens yes, because else the column is completely useless.

Do you need pointers ?

**Comment by noveens** [comment]
@PVince81, pointers would be a great way to start.

**Comment by PVince81** [comment]
Pointers

The job is executed here: https://github.com/owncloud/core/blob/master/cron.php#L122

I suggest you add statements around here https://github.com/owncloud/core/blob/master/lib/private/BackgroundJob/Job.php#L52 to measure the running time. Then store the value in the database.

For storing, use something like https://github.com/owncloud/core/blob/master/lib/private/BackgroundJob/JobList.php#L307. Maybe a new method `setLastRunTime()`.

**Comment by noveens** [comment]
@PVince81 , I have made the new method, but how to check the changes I have made?

File: lib/private/BackgroundJob/JobList.php [hunk 3]
```diff
@@ -311,4 +311,12 @@ public function setLastRun($job) {
 			->where($query->expr()->eq('id', $query->createNamedParameter($job->getId(), IQueryBuilder::PARAM_INT)));
 		$query->execute();
 	}
+
+	public function setExecutionTime($job, $timeTaken) {
+		$query = $this->connection->getQueryBuilder();
+		$query->update('jobs')
+			->set('last_run', $query->createNamedParameter($timeTaken, IQueryBuilder::PARAM_INT))
```
Comment (PVince81) [inline lib/private/BackgroundJob/JobList.php]: wrong column ?

**Comment by PVince81** [comment]
@noveens please have a look at my comment.

To test this you can run `php cron.php` manually.
If no job is running because of the schedule, you can reset the jobs manually with the following SQL query:
```
update oc_jobs set last_run=0,last_checked=0,reserved_at=0;
```

Then run cron.php again and see if your new column is populated.

File: lib/private/BackgroundJob/JobList.php (hunk 3 above)
Comment (noveens) [inline lib/private/BackgroundJob/JobList.php]: sorry, bad mistake :p

Comment (noveens) [inline lib/private/BackgroundJob/JobList.php]: @PVince81 , updated

**Comment by noveens** [comment]
@PVince81 , please recheck

**Comment by PVince81** [comment]
Code looks good so far. Thanks.

Will test it later.

**Comment by PVince81** [comment]
I tested it but after running cron.php I see the last_run being populated, but the value of "execution_duration" is always 0.

Even if my computer is too fast, I'd still expect values of at least a few millisecond.
Is that timestamp in seconds ?

Wanted to test it with LDAP (which provides ways to have slower jobs) but the app is broken at the moment: https://github.com/owncloud/core/issues/27154

**Comment by noveens** [comment]
@PVince81 ,
The new column I have added is of integer type since the time() function gives seconds since epoch
But the general case has time lesser than 1 second.

Hence I chose this approach.
If you need any updation, please tell me.

**Comment by PVince81** [comment]
- [ ] increase last digit in version.php to trigger an update

**Comment by PVince81** [comment]
Tested with LDAP. I saw cron.php taking a minute or so, still the column "execution_duration" contains only zeroes.

How did you test it?

**Comment by noveens** [comment]
@PVince81 ,
To be frank, I didn't test it because I couldn't find a way to populate the database.
Also, once populated how should I check the values in the database?

It would be great if you could tell me how to test this.

**Comment by PVince81** [comment]
Told you here: https://github.com/owncloud/core/pull/27144#issuecomment-279701127

The table is already pre-populated by default recurring jobs after setting up OC.

**Comment by noveens** [comment]
@PVince81 ,
Everything is fixed now, after lot of research and mind-breaking, my mistake was : 
I was calling setExecutionTime from the wrong class (had to be called from Job class rather than JobList class).

Please Recheck.

**Comment by PVince81** [comment]
Okay cool, it seems to work now.

I wonder if we should use millisecond values here to avoid having misleading zeroes in the column.

Thoughts ? @dercorn @jvillafanez @butonic @DeepDiver1975

**Comment by noveens** [comment]
@PVince81 ,
I think we can use gettimeofday( ) to get the execution time in the precision of even microseconds, though milli is better (considering overflow problems).
Should I change the code to store milliseconds rather than seconds?

**Comment by jvillafanez** [comment]
Please, include a log in the Job.php file. This will help to track the request id of the job and know what it's doing. Something like:
```
public function execute($jobList, ILogger $logger = null) {
  $jobList->setLastRun($this);
  try {
    $t1 = time();
    \OCP\Util::writeLog('jobs', 'Started background job ' . get_class($this) . print_r($this->argument, true), \OCP\Util::DEBUG);
      $this->run($this->argument);
    $secs = time() - $t1;
    \OCP\Util::writeLog('jobs', "Finished background job $secs seconds " . get_class($this) . print_r($this->argument, true), \OCP\Util::DEBUG);
  } catch (\Exception $e) {
```
So
* A log there is important. This will help to distinguish what log entries come from the background job. It's your choice if you want to wrap the `run` method in a "start" and "finish" logs.
* Logging the class with `get_class($this)` is important. This will allow to match the request id with the specific job, which will make easy to track down what the specific job is doing.
* Logging the arguments isn't really needed. I'm not sure if there is a job that uses a password, and not sure if the password will leak. In case of doubt, just ignore the arguments and stay safe.

**Comment by jvillafanez** [comment]
> I wonder if we should use millisecond values here to avoid having misleading zeroes in the column.

I don't think it's needed. If we need a default value to know what jobs haven't run already, just use null or -1, and let 0 be a valid value. We don't need such precision because we're interested in long running jobs.

**Comment by noveens** [comment]
@jvillafanez ,
I like the method of having a default value of -1.
I'll implement it.

I'll research more about the log part and get back if required.

**Comment by noveens** [comment]
@jvillafanez , I'll implement the log part as well

Updating changes in 10 mins.

**Comment by noveens** [comment]
@jvillafanez @PVince81 @DeepDiver1975 ,
Please review this PR since it seems it's okay to me now.

File: lib/private/BackgroundJob/Job.php [hunk 4]
```diff
@@ -49,7 +49,20 @@
 	public function execute($jobList, ILogger $logger = null) {
 		$jobList->setLastRun($this);
 		try {
+			//storing job start time
+			$jobStartTime = time();
+
+			\OCP\Util::writeLog('jobs', 'Started background job ' . get_class($this) . print_r($this->argument, true), \OCP\Util::DEBUG);
```
Comment (jvillafanez) [inline lib/private/BackgroundJob/Job.php]:
```
'Started background job ' . get_class($this) . ' : ' . json_encode($this->argument)
```
it seems a nicer message.

File: lib/private/BackgroundJob/Job.php [hunk 5]
```diff
@@ -49,7 +49,20 @@
...
+
 			$this->run($this->argument);
+
+			//storing job end time
+			$jobEndTime = time();
+			$timeTaken = $jobEndTime - $jobStartTime;
+
+			\OCP\Util::writeLog('jobs', "Finished background job $timeTaken seconds " . get_class($this) . print_r($this->argument, true), \OCP\Util::DEBUG);
```
Comment (jvillafanez) [inline lib/private/BackgroundJob/Job.php]:
```
'Finished background job ' . get_class($this) . ' : ' . json_encode($this->argument) . " (it took $timeTaken seconds)"
```
nicer message

**Comment by jvillafanez** [comment]
I don't have experience with migrations, but code looks good.

File: lib/private/BackgroundJob/Job.php (hunk 5 above)
Comment (PVince81) [inline lib/private/BackgroundJob/Job.php]: also change the app name to "cron", not "jobs".

**Comment by PVince81** [comment]
Apart from the cron message this looks good.

**Comment by noveens** [comment]
@bantu, @tanghus, @butonic, @DeepDiver1975 and @PVince81 ,

Please check the code, I think bug's solved now.

**Comment by jvillafanez** [comment]
```
{"reqId":"fpgDDP2MArnPXZMA\/Ni+","remoteAddr":"","app":"cron","message":"Run OCA\\Files_Sharing\\DeleteOrphanedSharesJob job with ID 12","level":0,"time":"2017-02-17T07:47:28+00:00","method":"--","url":"--","user":"--"}
{"reqId":"fpgDDP2MArnPXZMA\/Ni+","remoteAddr":"","app":"cron","message":"Finished OCA\\Files_Sharing\\DeleteOrphanedSharesJob job with ID 12","level":0,"time":"2017-02-17T07:47:28+00:00","method":"--","url":"--","user":"--"}
{"reqId":"fpgDDP2MArnPXZMA\/Ni+","remoteAddr":"","app":"cron","message":"Run OCA\\Files_Sharing\\ExpireSharesJob job with ID 13","level":0,"time":"2017-02-17T07:47:28+00:00","method":"--","url":"--","user":"--"}
{"reqId":"fpgDDP2MArnPXZMA\/Ni+","remoteAddr":"","app":"cron","message":"Finished OCA\\Files_Sharing\\ExpireSharesJob job with ID 13","level":0,"time":"2017-02-17T07:47:28+00:00","method":"--","url":"--","user":"--"}
{"reqId":"fpgDDP2MArnPXZMA\/Ni+","remoteAddr":"","app":"cron","message":"Run OC\\Authentication\\Token\\DefaultTokenCleanupJob job with ID 14","level":0,"time":"2017-02-17T07:47:28+00:00","method":"--","url":"--","user":"--"}
{"reqId":"fpgDDP2MArnPXZMA\/Ni+","remoteAddr":"","app":"cron","message":"Started background job of class : OC\\Authentication\\Token\\DefaultTokenCleanupJob","level":0,"time":"2017-02-17T07:47:28+00:00","method":"--","url":"--","user":"--"}
{"reqId":"fpgDDP2MArnPXZMA\/Ni+","remoteAddr":"","app":"no app in context","message":"Invalidating tokens older than 2017-02-16T07:47:28+00:00","level":1,"time":"2017-02-17T07:47:28+00:00","method":"--","url":"--","user":"--"}
{"reqId":"fpgDDP2MArnPXZMA\/Ni+","remoteAddr":"","app":"cron","message":"Finished background job, the job took : 0 seconds\njob is an instance of class : OC\\Authentication\\Token\\DefaultTokenCleanupJob","level":0,"time":"2017-02-17T07:47:28+00:00","method":"--","url":"--","user":"--"}
{"reqId":"fpgDDP2MArnPXZMA\/Ni+","remoteAddr":"","app":"cron","message":"Finished OC\\Authentication\\Token\\DefaultTokenCleanupJob job with ID 14","level":0,"time":"2017-02-17T07:47:28+00:00","method":"--","url":"--","user":"--"}
{"reqId":"fpgDDP2MArnPXZMA\/Ni+","remoteAddr":"","app":"cron","message":"Run OCA\\User_LDAP\\Jobs\\UpdateGroups job with ID 15","level":0,"time":"2017-02-17T07:47:28+00:00","method":"--","url":"--","user":"--"}
{"reqId":"fpgDDP2MArnPXZMA\/Ni+","remoteAddr":"","app":"cron","message":"Finished OCA\\User_LDAP\\Jobs\\UpdateGroups job with ID 15","level":0,"time":"2017-02-17T07:47:28+00:00","method":"--","url":"--","user":"--"}
```

A couple of things taking into account from the logs:
* The `PHP_EOL` should be removed. Each log entry will be in one line, so any newline char is an unwanted addition.
* You've put `get_class($this) . print_r($this->argument, true)`. There is no space between both the get class and the print_r, so the output might be confusing.
* For now, I haven't seen any output for the `print_r($this->argument, true)`. I'm not sure how the log will look like. @noveens if you've tested it, please show a couple of outputs: one with the print_r and another one with the json_encode, so we can check what format is more useful for the admin. (I'm not requesting any change for this at the moment).

Aside from this, we have another problem as you can see in the log output: duplicated and also false messages.

Those logs come from https://github.com/owncloud/core/blob/master/cron.php#L121 , and duplicate the new logs.

In addition, timed jobs can skip their own execution (https://github.com/owncloud/core/blob/master/lib/private/BackgroundJob/TimedJob.php#L52-L54), so the "Run blablabla job with ID 13" message could be a lie because the job can be skipped.

Taking this into account, I think we should include the removal of those message and stick with the new logs because they're more accurate.

**Comment by noveens** [comment]
@jvillafanez ,

As requested, I have deleted the log statements from :  https://github.com/owncloud/core/blob/master/cron.php#L121

Plus, Even I didn't find print_r($this->argument, true) print anything but IMO we should leave it so that if we have some arguments in future, it would print them.

File: lib/private/BackgroundJob/Job.php [hunk 6]
```diff
@@ -49,7 +49,24 @@
 	public function execute($jobList, ILogger $logger = null) {
 		$jobList->setLastRun($this);
 		try {
+			//storing job start time
+			$jobStartTime = time();
+
+			echo 'some' . PHP_EOL;
```
Comment (jvillafanez) [inline lib/private/BackgroundJob/Job.php]: remove this `echo` and the `gettype` below

**Comment by jvillafanez** [comment]
I think we're almost there. We'll have to check how it behaves and check if there is any additional unexpected consecuence.

From my point of view, just fix the comment above, and then it's ready for testing.

File: lib/private/BackgroundJob/Job.php (hunk 6 above)
Comment (noveens) [inline lib/private/BackgroundJob/Job.php]: sorry, forgot to remove the debug statements.

Updated changes pushed.

**Comment by jvillafanez** [comment]
@PVince81 @DeepDiver1975 I had to manually increase the OC version to force the trigger of the migration code (checked against daily master). I guess this will be fine once the code is merged. It's just to confirm that no more changes are needed.

Calling @owncloud/qa ....

Things to check ( @noveens feel free to add anything that has to be checked and I might be missing)
* New column "execution_time" in the "jobs" table with a default value of -1
* Running cron will:
    * set the execution time in seconds of the job in the "jobs" table. 0 is a valid value. It will help to identify what jobs have run and what jobs not.
    * add new logs entries in the owncloud.log. The logs entries should be clear enough to know what jobs have run, and should match the information contained in the "jobs" table.
* Need to test against a "heavy" system to verify we're counting the seconds right and not just setting zeros. Probably testing with LDAP active is enough to have some slow jobs which should take more than 1 second to execute.
* Find a job with arguments and check the log output.

Jenkins is complaining.... we'll need to check what happens with the tests.

**Comment by noveens** [comment]
@jvillafanez ,
the column name is "execution_duration", rest everything seems fine to me.

**Comment by noveens** [comment]
@bantu, @tanghus, @butonic, @DeepDiver1975 and @PVince81 ,

I managed to pass Jenkins now ( took 1h 42m :p ) please have a look and review this PR.

**Comment by jvillafanez** [comment]
Revert the permission change in the config.sample.php file please.

**Comment by davitol** [comment]
> New column "execution_time" in the "jobs" table with a default value of -1
> Running cron will:
> 
>     set the execution time in seconds of the job in the "jobs" table. 0 is a valid value. It will help to identify what jobs have run and what jobs not. ✅ 
>     add new logs entries in the owncloud.log. The logs entries should be clear enough to know what jobs have run, and should match the information contained in the "jobs" table.
> ✅ 
> Need to test against a "heavy" system to verify we're counting the seconds right and not just setting zeros. Probably testing with LDAP active is enough to have some slow jobs which should take more than 1 second to execute. ✅ 
> Find a job with arguments and check the log output.  ✅ 

:+1:

IMHO A small enhancement when running a job **without** arguments will be to not print the **with arguments : ** text


```
{"reqId":"4vkT2Z8Fy\/yzBApDPlqT","remoteAddr":"","app":"cron","message":"Finished background job, the job took : 0 seconds, this job is an instance of class : OCA\\Files\\BackgroundJob\\CleanupFileLocks with arguments : ","level":0,"time":"2017-02-20T12:08:15+00:00","method":"--","url":"--","user":"--"}
```

**Comment by noveens** [comment]
@jvillafanez ,
I didn't get the part I have to change,
you're talking about which permissions?

**Comment by jvillafanez** [comment]
screen shot 2017-02-21 at 09 05 54

That change in the permissions of the config.sample.php . It shouldn't appear.

**Comment by noveens** [comment]
@jvillafanez ,
pushed the changes.

**Comment by noveens** [comment]
@DeepDiver1975 ,
please re-review this

**Comment by jvillafanez** [comment]
I don't have anything more to add 👍

**Comment by noveens** [comment]
@PVince81 ,
Can this be merged, I think it's fine now.

**Comment by PVince81** [comment]
@DeepDiver1975 can you kick Jenkins to retest this ?

**Comment by noveens** [comment]
@PVince81 ,
This passed jenkins before my previous commit, which was just modifying the file permissions of config.sample.php which i suppose has nothing to do with any of the tests.

**Comment by PVince81** [comment]
Added to new features page: https://github.com/owncloud/core/wiki/ownCloud-10.0-Features

**Comment by jvillafanez** [comment]
@PVince81 @DeepDiver1975 Backport the logging part? I think it could be quite useful and not intrusive. DB changes won't be backported.

**Comment by PVince81** [comment]
Yes, please backport the logging part.

@noveens can you do it ? Basically create a branch off "stable9.1", put your commits there then submit a PR that targets stable9.1. Only put the logging part there.

**Comment by noveens** [comment]
@PVince81 ,
Sure I can do that.
Just the logging part right?
And do I have to update anything in this branch?

**Comment by PVince81** [comment]
Nothing to update here. Make a new branch.

**Comment by PVince81** [comment]
stable9.1 logging changes backport: https://github.com/owncloud/core/pull/27239

**Pull request merged**